### PR TITLE
Added a Problem Details Provider to be able to return the ProblemDetails from an error

### PR DESCRIPTION
**Pull request by DanielToft**
Hi

Thank's for your great package!

In my project i need to create an ProblemDetails object from an Exception, to support SignalR. To be able to do that i need an public expose method to return the ProblemDetails. I have created a ProblemDetails provider to support that. 

I have not tested it, because i'm unable to build the project. I hobe you will accept my pull request.

### CODE REVIEW

**Comment by khellang** [comment]
> In my project i need to create an ProblemDetails object from an Exception, to support SignalR.

Is there a reason you can't use the existing mapping from an exception to a problem details object?

**Comment by DanielToft** [comment]
I'm using the existing mapping, but that's not the problem. I need a way to get the right Problem Details object, based on my exception to problem details mapping. Today it's an internal method.

Like this:

```
try {
     throw new CustomException();
}
catch(Exception e)
{
    var problem = GetProblemDetailsFromMap(e);
   
    // And then do something with the problem object
}
```

**Comment by khellang** [comment]
I guess I'm just not seeing how that's different from using

```csharp
options.Map<Exception>(e =>
{
    var problem = GetProblemDetailsFromMap(e);

    // And then do something with the problem object
});
```

**Comment by DanielToft** [comment]
I don't think you understand what I am writing. Have you looked at my committed code?

It's not about mapping, it's about retrieve the problem object based on the mapping.

I'm using your package in my project and it works great for my controllers. However, in SignalR it's not working because they have a different implementation. So to be able to return a ProblemDetails object to my clients, with the same format and configuration i have configured with your package, a need a method to return the problem. 

In my pull request i have created a public class "ProblemDetailsProvider" (maybe it should be named differently), and moved the GetDetails method into that and marked it public.

Then i will use the ProblemDetailsProvider like this:

```csharp
public class MyHub : Hub
{
    private readonly ProblemDetailsProvider _problemDetailsProvider;

    public MyHub(ProblemDetailsProvider problemDetailsProvider)
    {
        _problemDetailsProvider = problemDetailsProvider;
    }

    [HubMethodName("Test")]
    public async Task<object> Test(string id)
    {
        try
        {
            return await service.DoStuff();
        }
        catch (Exception e)
        {
            return new SignalRErrorResponse() {
                Problem = _problemDetailsProvider.GetDetails(context, e)
            };
        }
    }
}
```

**Comment by khellang** [comment]
> I don't think you understand what I am writing. Have you looked at my committed code?

No, I really don't 😝 Yes, I have looked at the code, but you included no sample (and no tests) to show how you intended to use this "provider", so you haven't really made it easy for me 😉

So basically you just want to pull out the mapping code from the middleware into a separate service so you can utilize the same mapping in other places?

**Comment by DanielToft** [comment]
I'm sorry! :disappointed:

And yes, it's exactly what i want to be able to do.

File: src/ProblemDetails/ProblemDetailsMiddleware.cs [hunk 1]
```diff
@@ -83,9 +73,9 @@ public async Task Invoke(HttpContext context)
                 {
                     ClearResponse(context, StatusCodes.Status500InternalServerError);
 
-                    var details = GetDetails(context, error);
+                    var details = _problemDetailsProvider.GetDetails(context, error);
 
-                    if (Options.ShouldLogUnhandledException(context, error, details))
+                    if (_problemDetailsProvider.Options.ShouldLogUnhandledException(context, error, details))
```
Comment (khellang) [inline src/ProblemDetails/ProblemDetailsMiddleware.cs]: I'm not a big fan of accessing the options through this provider, can we just make the provider be a `Func<HttpContext, Exception, MvcProblemDetails>` on the options instead?

Comment (DanielToft) [reply src/ProblemDetails/ProblemDetailsMiddleware.cs]: Yes that's fine.

**Comment by khellang** [comment]
Hey @DanielToft! Since this change has been sitting here for a while, it has become quite stale. I did add a `ProblemDetailsFactory` to a recent version of the middleware if you want to test it out 😄

**Comment by DanielToft** [comment]
Hey @khellang 

Thank you for the implementation! It looks really nice and seems to do what i need. I'll test it later, and reach out to you if i have any problems. You can close this PR. 

Thank you!

**Comment by khellang** [comment]
> You can close this PR.

I'm sorry I just went ahead and added this without pinging you first. I got a bit carried away 😅

**Pull request closed without merging**